### PR TITLE
feat: upgrade to go 1.18 and support modules via workspaces

### PR DESCRIPTION
**Pull request by LucasRoesler**
## Description

Support Go 1.18 and improve module support using Workspaces

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## Which issue(s) this PR fixes 

Fixes #

## How Has This Been Tested?

See https://github.com/LucasRoesler/openfaas-golang-template-workspace

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [?] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Version change (see: Impact to existing users)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Examples that come to my mind:

* Private Go module usage, perhaps with vendoring
* Named sub-packages i.e. routes, imported into the handler
* Normal Go modules with external packages
* Normal Go modules on, with public vendored packages?

File: template/go/Dockerfile [hunk 1]
```diff
@@ -24,13 +24,11 @@ ENV CGO_ENABLED=0
...
 COPY . .
 
-# Add user overrides to the root go.mod, which is the only place "replace" can be used
-RUN cat function/GO_REPLACE.txt >> ./go.mod || exit 0
-
 # Run a gofmt and exclude all vendored code.
 RUN test -z "$(gofmt -l $(find . -type f -name '*.go' -not -path "./vendor/*" -not -path "./function/vendor/*"))" || { echo "Run \"gofmt -s -w\" on your Golang code"; exit 1; }
 
 WORKDIR /go/src/handler/function
+RUN go env
```
Comment (alexellis) [inline template/go/Dockerfile:31]: Would we remove this debug?

**Review by alexellis — APPROVED** [review]
Looks great to me :raised_hands:

**Comment by LucasRoesler** [comment]
@alexellis i remove the `go env` statement and i have updated the readme here https://github.com/LucasRoesler/openfaas-golang-template-workspace which explains the supported modes, please make sure to review that before we merge.

**Review by alexellis — APPROVED** [review]
Approved for merge

**Comment by alexellis** [comment]
I took a look at the examples repo, the only thing that would make them better is dropping a CI build into the repo to validate them automatically.

I have an example you could [adapt in this blog post](https://www.openfaas.com/blog/edge-actions/)

**Pull request merged**